### PR TITLE
Submit equivocation reports for Aura and Babe

### PR DESCRIPTION
**Pull request by marcio-diaz**
We need	to submit unsigned transactions reporting equivocations in Aura and Babe.

When detecting an equivocation using `check_header` we submit an unsigned transaction,
providing the equivocating headers, to the
transaction pool in a dummy way	(without taking into account nounces, how to do it right?).
Then we	implement `ValidateUnsigned` to	check that the equivocation proof is correct.
The slashing part should be implemented	later.

Changes:

- Move `digest.rs` and `find_pre_digest` to `aura/primitives/src/digest`, to be used in the runtime.
- Implement `EquivocationProof` for Aura and Babe.
- Move `slot_author` to `aura/primitives`, to be used in the runtime. (There should be a better way...)
- Add `transaction_pool` to `check_header`, to be able to submit unsigned tx reporting equivocations.
- Add `transaction_queue` to `import_queue` in Aura.
- Create `/consensus/safety` module.
- Add dummy `submit_report_call` and `EquivocationProof` trait to safety module.
- Add `Option<transaction_pool>` to `build_import_queue`.
- Introduce new types for Aura trait in Runtime: `Signature, DigestItem, Digest, Header`.
- Implement `ValidateUnsigned` for Aura, adding a function to check equivocations.

TODO:

- Implement `ValidateUnsigned` for Babe.
- Implement the real `submit_report_call`, taking in account nounce and other stuff (?).
- Add tests.

Questions:
- Is there an API to submit unsigned transactions in a proper way?

### CODE REVIEW

File: srml/babe/src/lib.rs [hunk 1]
```diff
@@ -111,7 +111,36 @@ decl_storage! {
...
+		match call {
+			Call::report_equivocation(proof) => {
+				let maybe_equivocation_proof = EquivocationProof::decode(proof);
+				if let Some(equivocation_proof) = maybe_equivocation_proof {
+					if equivocation_proof.is_valid() {
+						return TransactionValidity::Valid {
+							priority: 0,
+							requires: vec![],
+							provides: vec![],
+							longevity: std::u64::MAX,
```
Comment (bkchr) [inline srml/babe/src/lib.rs]: Same

File: srml/executive/src/lib.rs [hunk 2]
```diff
@@ -316,6 +316,8 @@ where
 		const MISSING_SENDER: i8 = -20;
 		const INVALID_INDEX: i8 = -10;
 
+		println!("running validate_transaction");
```
Comment (bkchr) [inline srml/executive/src/lib.rs]: Printlns

File: core/consensus/aura/primitives/Cargo.toml [hunk 3]
```diff
@@ -6,8 +6,11 @@ description = "Primitives for Aura consensus"
 edition = "2018"
 
 [dependencies]
+parity-codec = "3.4"
```
[the remark this answers is not in the document]
Comment (DemiMarie-parity) [reply core/consensus/aura/primitives/Cargo.toml]: Make that 4.0.0

**Comment by DemiMarie-parity** [comment]
@marcio-diaz Is this superceded by #2800?  If so, it should be closed.

**Pull request closed without merging**